cl/redisapp/retry.go

```diff
@@ -0,0 +1,113 @@
+package redisapp
+
+import (
+	"context"
+	"errors"
+	"log"
+	"math"
+	"sync"
+	"time"
+)
+
+const (
+	initialBackoff = 200 * time.Millisecond // Initial backoff duration
+	maxBackoff     = 30 * time.Second       // Maximum backoff duration
+)
+
+var (
+	retryMutex              sync.Mutex
+	ErrFailedAfterNAttempts = errors.New("operation failed after N attempts")
+)
+
+// Backoff function implementing an exponential backoff strategy
+func backoff(attempt int) time.Duration {
+	backoff := float64(initialBackoff) * math.Pow(2, float64(attempt))
+	if backoff > float64(maxBackoff) {
+		backoff = float64(maxBackoff)
+	}
+	return time.Duration(backoff)
+}
+
+func retryWithInfiniteBackoff(ctx context.Context, operation func() (bool, error)) (bool, error) {
+	retryMutex.Lock()
```

Comment:
If the redis client doesnt provide synchronization, this external mutex doesnt really help in my opinion. If it does not, then there should be synchronization at the package level where it is used.